File: src/core/Shared/Adapters/CSV/CSVExport.cs
using System;
using System.Collections.Generic;
using System.Linq;
using core.Account;
using core.Cryptos;
using core.Notes;
using core.Options;
using core.Portfolio;
using core.Stocks;
using core.Stocks.Services.Trading;
using PositionInstance = core.Stocks.PositionInstance;

namespace core.Shared.Adapters.CSV
{
    public class CSVExport
    {
        private const string DATE_FORMAT = "yyyy-MM-dd";
        private record struct PendingPositionRecord(string ticker, decimal bid, decimal numberOfShares, decimal? stopPrice, string date, string closed, bool purchased, string strategy, string notes);
        private record struct StockRecord(string ticker, string type, decimal amount, decimal price, string date, string notes);
        private record struct NoteRecord(string created, string ticker, string note);
        private record struct OptionRecord(string ticker, string type, decimal strike, string optiontype, string expiration, decimal amount, decimal premium, string filled);
        private record struct UserRecord(string email, string first_name, string last_name);
        private record struct CryptosRecord(string symbol, string type, decimal amount, decimal price, string date);
        private record struct TradesRecord(string symbol, string opened, string closed, decimal daysheld, decimal firstbuycost, decimal cost, decimal profit, decimal returnpct, decimal rr, decimal? riskedAmount, string grade, string gradeNote);
        private record struct StockListRecord(string ticker, string created, string notes);
        private record struct StockListRecordJustTicker(string ticker);
        private record struct TradingStrategyResultRecord(string strategyName, string ticker, decimal profit, decimal rr, decimal returnPct, decimal numberOfShares, decimal cost, decimal averageBuyCostPerShare, decimal averageSaleCostPerShare, string opened, string closed, decimal daysHeld);

        public static string Generate(ICSVWriter writer, List<TradingStrategyPerformance> strategies)
        {
            var rows = strategies
                .SelectMany(strategy => 
                    strategy.positions.Select(r =>
                        new TradingStrategyResultRecord(
                            strategy.strategyName,
                            r.Ticker,
                            Math.Round(r.Profit, 2),
                            Math.Round(r.RR, 2),
                            Math.Round(r.GainPct, 2) * 100,
                            r.CompletedPositionShares,
                            r.CompletedPositionShares * r.CompletedPositionCostPerShare,
                            r.CompletedPositionCostPerShare,
                            r.AverageSaleCostPerShare,
                            r.Opened.Value.ToString(DATE_FORMAT),
                            r.Closed.Value.ToString(DATE_FORMAT),
                            r.DaysHeld
                        )
                    )
                );

            return writer.Generate(rows);
        }

        public static string Generate(ICSVWriter writer, IEnumerable<PendingStockPosition> pending)
        {
            var rows = pending.Select(p => p.State).Select(p => new PendingPositionRecord(
                ticker: p.Ticker,
                bid: p.Bid,
                numberOfShares: p.NumberOfShares,
                stopPrice: p.StopPrice,
                date: p.Date.ToString(DATE_FORMAT),
                closed: p.Closed?.ToString(DATE_FORMAT),
                purchased: p.Purchased,
                strategy: p.Strategy,
                notes: p.Notes
            ));

            return writer.Generate(rows);
        }

        public static string Generate(ICSVWriter writer, StockListState list, bool justTickers)
        {
            return justTickers switch {
                true => writer.Generate(list.Tickers.Select(s => new StockListRecordJustTicker(s.Ticker))),
                false => writer.Generate(list.Tickers.Select(s => new StockListRecord(s.Ticker, s.When.ToString(DATE_FORMAT), s.Note)))
            };
        }

        public static string Generate(ICSVWriter writer, IEnumerable<PositionInstance> trades)
        {
            var rows = trades.Select(t =>
                new TradesRecord(
                    symbol: t.Ticker,
                    opened: t.Opened?.ToString(DATE_FORMAT),
                    closed: t.Closed?.ToString(DATE_FORMAT),
                    daysheld: t.DaysHeld,
                    firstbuycost: t.CompletedPositionCostPerShare,
                    cost: t.Cost,
                    profit: t.Profit,
                    returnpct: t.IsClosed ? t.GainPct : t.UnrealizedGainPct.Value,
                    rr: t.RR,
                    riskedAmount: t.RiskedAmount,
                    grade: t.Grade,
                    gradeNote: t.GradeNote
                )
            );

            return writer.Generate(rows);
        }
        public static string Generate(ICSVWriter writer, IEnumerable<User> users)
        {
            var rows = users.Select(u => new UserRecord(u.State.Email, u.State.Firstname, u.State.Lastname));

            return writer.Generate(rows);
        }
        
        public static string Generate(ICSVWriter writer, IEnumerable<OwnedCrypto> cryptos)
        {
            var rows = cryptos.SelectMany(o => o.State.UndeletedBuysOrSells.Select(e => (o, e)))
                .OrderBy(t => t.e.When)
                .Select(t => t.e switch {
                    CryptoPurchased cp => new CryptosRecord(t.o.State.Token, "buy", cp.Quantity, cp.DollarAmount, cp.When.ToString(DATE_FORMAT)),
                    CryptoSold cs => new CryptosRecord(t.o.State.Token, "sell", cs.Quantity, cs.DollarAmount, cs.When.ToString(DATE_FORMAT)),
                    _ => new CryptosRecord()
                })
                .Where(r => r.symbol != null);

            return writer.Generate(rows);
        }

        public static string Generate(ICSVWriter writer, IEnumerable<OwnedStock> stocks)
        {
            var rows = stocks
                .SelectMany(o => o.State.BuyOrSell)
                .Select(e => 
                    e switch {
                        StockPurchased sp => new StockRecord(sp.Ticker, "buy", sp.NumberOfShares, sp.Price, sp.When.ToString(DATE_FORMAT), sp.Notes),
                        StockPurchased_v2 sp => new StockRecord(sp.Ticker, "buy", sp.NumberOfShares, sp.Price, sp.When.ToString(DATE_FORMAT), sp.Notes),
                        StockSold ss => new StockRecord(ss.Ticker, "sell", ss.NumberOfShares, ss.Price, ss.When.ToString(DATE_FORMAT), ss.Notes),
                        _ => new StockRecord()
                    }
                )
                .Where(s => s.ticker != null);

            return writer.Generate(rows);
        }

        private static IEnumerable<object> StockEventToParts((Aggregate a, AggregateEvent ae) tuple)
        {
            switch(tuple.ae)
            {
                case CryptoPurchased cp:

                    return new object[] {
                        cp.Token,
                        "buy",
                        cp.Quantity,
                        cp.DollarAmount,
                        cp.When.ToString(DATE_FORMAT)
                    };

                case CryptoSold cs:

                    return new object[] {
                        cs.Token,
                        "sell",
                        cs.Quantity,
                        cs.DollarAmount,
                        cs.When.ToString(DATE_FORMAT)
                    };

                default:
                    throw new InvalidOperationException("Invalid stock event for export: " + tuple.ae.GetType());
            }
        }

        public static string Generate(ICSVWriter writer, IEnumerable<Note> notes)
        {
            var rows = notes.Select(n => {
                var state = n.State;

                return new NoteRecord(
                    state.Created.ToString(DATE_FORMAT),
                    state.RelatedToTicker,
                    state.Note
                );
            });

            return writer.Generate(rows);
        }

        public static string Generate(ICSVWriter writer, IEnumerable<OwnedOption> options)
        {
            var rows = options.SelectMany(o => o.State.Buys.Select(op => (o, (AggregateEvent)op)))
                .Union(options.SelectMany(o => o.State.Sells.Select(os => (o, (AggregateEvent)os))))
                .Union(options.SelectMany(o => o.State.Expirations.Select(os => (o, (AggregateEvent)os))))
                .OrderBy(t => t.Item2.When)
                .Select(t => t.Item2 switch {
                    OptionSold s => new OptionRecord(
                        t.o.State.Ticker,
                        "sell",
                        t.o.State.StrikePrice,
                        t.o.State.OptionType.ToString(),
                        t.o.State.Expiration.ToString(DATE_FORMAT),
                        s.NumberOfContracts,
                        s.Premium,
                        s.When.ToString(DATE_FORMAT)
                    ),
                    OptionPurchased p => new OptionRecord(
                        t.o.State.Ticker,
                        "buy",
                        t.o.State.StrikePrice,
                        t.o.State.OptionType.ToString(),
                        t.o.State.Expiration.ToString(DATE_FORMAT),
                        p.NumberOfContracts,
                        p.Premium,
                        p.When.ToString(DATE_FORMAT)
                    ),
                    OptionExpired ex => new OptionRecord(
                        t.o.State.Ticker,
                        ex.Assigned ? "assigned" : "expired",
                        t.o.State.StrikePrice,
                        t.o.State.OptionType.ToString(),
                        t.o.State.Expiration.ToString(DATE_FORMAT),
                        0,
                        0,
                        ex.When.ToString(DATE_FORMAT)
                    ),
                    _ => new OptionRecord()
                })
                .Where(s => s.ticker != null);

            return writer.Generate(rows);
        }

        public static string GenerateFilename(string prefix)
        {
            return prefix + "_" + DateTime.UtcNow.ToString($"yyyyMMdd_hhmss") + ".csv";
        }
    }
}

File: src/core.fs/Services/CSVExport.fs
namespace core.fs.Services

open System
open core.Cryptos
open core.Stocks
open core.fs.Accounts
open core.fs.Adapters.CSV
open core.fs.Options
open core.fs.Services.Trading
open core.fs.Stocks

module CSVExport =
    
    [<Literal>]
    let private DATE_FORMAT = "yyyy-MM-dd"
    [<Literal>]
    let private CURRENCY_FORMAT = "C"
    [<Literal>]
    let private PERCENT_FORMAT = "P"
    [<Literal>]
    let private NUMBER_FORMAT = "N2"
    
    let private number culture (value:decimal) = value.ToString(NUMBER_FORMAT, culture)
    let private numberOption culture (opt:decimal option) = opt |> Option.map (number culture) |> Option.defaultValue ""
    let private currency culture (value:decimal) = value.ToString(CURRENCY_FORMAT, culture)
    let private percent culture (value:decimal) = value.ToString(PERCENT_FORMAT, culture)
    let private currencyOption culture (opt:decimal option) = opt |> Option.map (currency culture) |> Option.defaultValue ""
    let private date (value:DateTimeOffset) = value.ToString(DATE_FORMAT)
    let private dateOption (opt:DateTimeOffset option) = opt |> Option.map date |> Option.defaultValue ""
    
    type PendingPositionRecord =
        {
            Ticker:string
            Bid:decimal
            NumberOfShares:decimal
            StopPrice:string
            Created:string
            Closed:string
            Purchased:bool
            Strategy:string
            CloseReason:string
            OrderType:string
            OrderDuration:string
            Notes:string
        }
        
    type StockRecord =
        {
            Ticker:string
            Type:string
            Amount:decimal
            Price:decimal
            Date:string
        }
        
    type NoteRecord =
        {
            Created:string
            Ticker:string
            Note:string
        }
        
    type OptionRecord =
        {
            Ticker:string
            Type:string
            Strike:decimal
            OptionType:string
            Expiration:string
            Amount:decimal
            Premium:decimal
            Filled:string
        }
        
    type UserRecord =
        {
            Email:string
            Firstname:string
            Lastname:string
        }
        
    type CryptosRecord =
        {
            Symbol:string
            Type:string
            Amount:decimal
            Price:decimal
            Date:string
        }
    
    type TradesRecord =
        {
            StrategyName:string
            Ticker:string
            PositionType:string
            NumberOfShares:string
            Opened:string
            Closed:string
            DaysHeld:int
            AverageCostPerShare:string
            LastSellCostPerShare:string
            AverageBuyCostPerShare:string
            AverageSaleCostPerShare:string
            CompletedPositionCostPerShare:string
            StopPrice:string
            Cost:string
            Profit:string
            ReturnPct:string
            RR:string
            InitialRiskedAmount:string
            RiskAmount:string
            MaxDrawdown:string
            MaxGain:string
            MaxDrawdownFirst10Bars:string
            MaxGainFirst10Bars:string
            Strategy:string
            Grade:string
            EntryNote:string
            GradeNote:string
        }
        
    type StockListRecord =
        {
            Ticker:string
            Created:string
            Notes:string
        }
        
    type StockListRecordJustTicker =
        {
            Ticker:string
        }
        
    let private mapToTradeRecord culture strategyName maxDrawdown maxGain maxDrawdownFirst10Bars maxGainFirst10Bars (t:StockPositionWithCalculations) =
        {
            StrategyName = strategyName
            Ticker = t.Ticker.Value
            PositionType = t.StockPositionType
            NumberOfShares = (if t.IsClosed then t.CompletedPositionShares else t.NumberOfShares) |> number culture
            Opened = t.Opened |> date
            Closed = t.Closed |> dateOption
            DaysHeld = t.DaysHeld
            AverageBuyCostPerShare = t.AverageBuyCostPerShare |> number culture
            AverageSaleCostPerShare = t.AverageSaleCostPerShare |> number culture 
            AverageCostPerShare = t.CompletedPositionCostPerShare |> number culture
            LastSellCostPerShare = t.ClosePrice |> numberOption culture
            CompletedPositionCostPerShare = t.CompletedPositionCostPerShare |> number culture
            StopPrice = t.StopPrice  |> numberOption culture
            Cost = t.Cost |> number culture
            Profit = t.Profit |> number culture
            ReturnPct = t.GainPct |> percent culture
            RR = t.RR |> number culture
            InitialRiskedAmount = t.InitialRiskedAmount |> numberOption culture
            RiskAmount = t.RiskedAmount |> numberOption culture
            MaxDrawdown = maxDrawdown |> numberOption culture
            MaxGain = maxGain |> numberOption culture
            MaxDrawdownFirst10Bars = maxDrawdownFirst10Bars |> numberOption culture
            MaxGainFirst10Bars = maxGainFirst10Bars |> numberOption culture
            Strategy = t.TryGetLabelValue("strategy") |> Option.defaultValue ""
            Grade = if t.Grade.IsSome then t.Grade.Value.Value else ""
            GradeNote = t.GradeNote |> Option.defaultValue ""
            EntryNote = t.Notes |> List.tryHead |> Option.map(_.content) |> Option.defaultValue ""
        }
        
    let strategyPerformance (culture:IFormatProvider) (writer:ICSVWriter) (strategies:seq<TradingStrategyPerformance>) =
        
        let rows =
            strategies
            |> Seq.collect (fun strategy -> 
                strategy.results
                |> Seq.map (fun result ->
                    let maxDrawdown = result.MaxDrawdownPct |> Some
                    let maxGain = result.MaxGainPct |> Some
                    let maxDrawdownFirst10Bars = result.MaxDrawdownFirst10Bars |> Some
                    let maxGainFirst10Bars = result.MaxGainFirst10Bars |> Some
                    mapToTradeRecord culture strategy.strategyName maxDrawdown maxGain maxDrawdownFirst10Bars maxGainFirst10Bars result.Position)
            )
            
        writer.Generate(rows)
    
    let pendingPositions (culture:IFormatProvider) (writer:ICSVWriter) (pending:seq<PendingStockPosition>) =
        
        let rows =
            pending
            |> Seq.map (fun p ->
                {
                    Ticker = p.State.Ticker.Value
                    Bid = p.State.Bid
                    NumberOfShares = p.State.NumberOfShares
                    StopPrice = p.State.StopPrice |> numberOption culture
                    Created = p.State.Created |> date
                    Closed = p.State.Closed |> dateOption
                    Purchased = p.State.Purchased
                    Strategy = p.State.Strategy
                    Notes = p.State.Notes
                    CloseReason = p.State.CloseReason
                    OrderDuration = p.State.OrderDuration
                    OrderType = p.State.OrderType 
                }
            )
            
        writer.Generate(rows)
        
    let stockList (writer:ICSVWriter) (list:StockListState) (justTickers:bool) =
        
        match justTickers with
        | true -> writer.Generate(list.Tickers |> Seq.map (fun s -> { Ticker = s.Ticker.Value }))
        | false -> writer.Generate(list.Tickers |> Seq.map (fun s -> { Ticker = s.Ticker.Value; Created = s.When |> date; Notes = s.Note }))
        
    
    let trades culture (writer:ICSVWriter) (trades:seq<StockPositionWithCalculations>) =
        
        let rows =
            trades
            |> Seq.map (fun p ->
                let mae = p.TryGetLabelValue("mae") |> Option.map decimal
                let mfe = p.TryGetLabelValue("mfe") |> Option.map decimal
                let maeFirst10Bars = p.TryGetLabelValue("mae10") |> Option.map decimal
                let mfeFirst10Bars = p.TryGetLabelValue("mfe10") |> Option.map decimal
                
                mapToTradeRecord culture "Actual Trade" mae mfe maeFirst10Bars mfeFirst10Bars p
            )
            
        writer.Generate(rows)
        
    let users (writer:ICSVWriter) (users:seq<User>) =
        
        let rows =
            users
            |> Seq.map (fun u ->
                {
                    Email = u.State.Email
                    Firstname = u.State.Firstname
                    Lastname = u.State.Lastname
                }
            )
            
        writer.Generate(rows)
        
    let cryptos (writer:ICSVWriter) (cryptos:seq<OwnedCrypto>) =
        
        let rows =
            cryptos
            |> Seq.collect (fun o -> o.State.UndeletedBuysOrSells |> Seq.map (fun e -> (o, e)))
            |> Seq.sortBy (fun (_, e) -> e.When)
            |> Seq.map (fun (o, e) ->
                match e with
                | :? CryptoPurchased as cp -> 
                    {
                        Symbol = o.State.Token
                        Type = "buy"
                        Amount = cp.Quantity
                        Price = cp.DollarAmount
                        Date = cp.When |> date
                    }
                | :? CryptoSold as cs ->
                    {
                        Symbol = o.State.Token
                        Type = "sell"
                        Amount = cs.Quantity
                        Price = cs.DollarAmount
                        Date = cs.When |> date
                    }
                | _ -> { Symbol = ""; Type = ""; Amount = 0m; Price = 0m; Date = "" }
            )
            |> Seq.filter (fun r -> r.Symbol <> "")
            
        writer.Generate(rows)
        
    let stocks (writer:ICSVWriter) (stocks:seq<StockPositionState>) =
        
        let rows =
            stocks
            |> Seq.collect (fun o -> o.ShareTransactions |> Seq.map (fun e -> (o, e)))
            |> Seq.sortBy (fun (_, e) -> e.Date)
            |> Seq.map (fun (o, e) ->
                {
                    Ticker = o.Ticker.Value
                    Type = e.Type.ToString().ToLower()
                    Amount = e.NumberOfShares
                    Price = e.Price
                    Date = e.Date |> date
                }
            )
            |> Seq.filter (fun r -> r.Ticker <> "")
            
        writer.Generate(rows)
        
    let options (writer:ICSVWriter) (options:seq<OptionPositionState>) =
        
        let data = options |> Seq.collect (fun o -> o.Events |> Seq.map (fun e -> (o, e)))
            
        let rows =
            data
            |> Seq.sortBy (fun (_, e) -> e.When)
            |> Seq.map (fun (o, e) -> 
                match e with
                | :? OptionContractSoldToOpen as os ->
                    {
                        Ticker = o.UnderlyingTicker.Value
                        Type = "sell"
                        Strike = os.Strike
                        OptionType = os.OptionType
                        Expiration = os.Expiration
                        Amount = os.Quantity |> decimal
                        Premium = os.Price
                        Filled = os.When |> date
                    }
                | :? OptionContractBoughtToOpen as op ->
                    {
                        Ticker = o.UnderlyingTicker.Value
                        Type = "buy"
                        Strike = op.Strike
                        OptionType = op.OptionType
                        Expiration = op.Expiration
                        Amount = op.Quantity |> decimal
                        Premium = op.Price
                        Filled = op.When |> date
                    }
                | :? OptionContractBoughtToClose as bc ->
                    {
                        Ticker = o.UnderlyingTicker.Value
                        Type = "buy"
                        Strike = bc.Strike
                        OptionType = bc.OptionType
                        Expiration = bc.Expiration
                        Amount = bc.Quantity |> decimal
                        Premium = bc.Price
                        Filled = bc.When |> date
                    }
                | :? OptionContractSoldToClose as sc ->
                    {
                        Ticker = o.UnderlyingTicker.Value
                        Type = "sell"
                        Strike = sc.Strike
                        OptionType = sc.OptionType
                        Expiration = sc.Expiration
                        Amount = sc.Quantity |> decimal
                        Premium = sc.Price
                        Filled = sc.When |> date
                    }
                | _ -> { Ticker = ""; Type = ""; Strike = 0m; OptionType = ""; Expiration = ""; Amount = 0m; Premium = 0m; Filled = "" }
            )
            |> Seq.filter (fun r -> r.Ticker <> "")
            
        writer.Generate(rows)


    let generateFilename prefix =
        prefix + "_" + DateTime.UtcNow.ToString("yyyyMMdd_hhmss") + ".csv"


